reporter/gcppubsub/pubsub.go

```diff
@@ -0,0 +1,109 @@
+package gcppubsub
+
+import (
+	"cloud.google.com/go/pubsub"
+	"context"
+	"encoding/json"
+	"github.com/openzipkin/zipkin-go/model"
+	"github.com/openzipkin/zipkin-go/reporter"
+	"log"
+	"os"
+)
+
+const defaultPubSubTopic = "defaultTopic"
+
+// Reporter implements Reporter by publishing spans to a GCP gcppubsub.
+type Reporter struct {
+	logger *log.Logger
+	topic  string
+	client *pubsub.Client
+}
+
+// ReporterOption sets a parameter for the PubSubReporter
+type ReporterOption func(c *Reporter)
+
+// Send send span to topic
+func (r *Reporter) Send(s model.SpanModel) {
+	// Zipkin expects the message to be wrapped in an array
+	ss := []model.SpanModel{s}
+	m, err := json.Marshal(ss)
+	if err != nil {
+		r.logger.Printf("failed when marshalling the span: %s\n", err.Error())
+		return
+	}
+	err = r.publish(m)
+	if err != nil {
+		r.logger.Printf("Error publishing message to gcppubsub: %s msg: %s", err.Error(), string(m))
+	}
+}
+
+// Close close span
+func (r *Reporter) Close() error {
+	return r.client.Close()
+}
+
+// Logger sets the logger used to report errors in the collection
+// process.
+func Logger(logger *log.Logger) ReporterOption {
+	return func(c *Reporter) {
+		c.logger = logger
+	}
+}
+
+// Client sets the client used to produce to gcppubsub.
+func Client(client *pubsub.Client) ReporterOption {
+	return func(c *Reporter) {
+		c.client = client
+	}
+}
+
+// Topic sets the gcppubsub topic to attach the reporter producer on.
+func Topic(t string) ReporterOption {
+	return func(c *Reporter) {
+		c.topic = t
+	}
+}
+
+// NewReporter returns a new gcppubsub-backed Reporter. address should be a slice of
+// TCP endpoints of the form "host:port".
+func NewReporter(options ...ReporterOption) (reporter.Reporter, error) {
+	topic := os.Getenv("ZIPKIN_PUBSUB_TOPIC")
```

Comment:
I am curious on why to pass this as an env var over passing this as an argument.

Reply:
May be is cleaner to have topic name as parameter instead of env var. Regarding others env vars... I think it is the best alternative a most probably gcp pubsub reporter will be executed in GCP... where credentials and project env variables are automatically set in GKE or cloud functions... 

I will update the PR making topic name as parameter...